
# MICHIGAN IMPUTATION SERVER

## Prepare data for imputation

Install plink


```R
install plink:
wget https://www.cog-genomics.org/static/bin/plink180612/plink_linux_x86_64.zip
unzip plink_linux_x86_64.zip
chmod +x plink
plink
```

Script to prepare plink data for the Michigan Imputation Server obtained from: https://github.com/isglobal-brge/brgeUtils/blob/master/impute_regions/scriptPreproc.R


```R
plink --bfile mach_learn_alz1 --recode --out mach_learn_alz

sed 's/^23/X/' mach_learn_alz.map  > mach_learn18.map 
cp mach_learn_alz.ped mach_learn18.ped


# Convert hg18 to hg19
python liftOverPlink.py --map mach_learn18.map --out lifted --chain hg18ToHg19.over.chain.gz --bin liftOver
python rmBadLifts.py --map lifted.map --out good_lifted.map --log bad_lifted.dat

cut -f 2 bad_lifted.dat > to_exclude.dat
cut -f 4 liftelifd.bed.unlifted | sed "/^#/d" >> to_exclude.dat
plink --file mach_learn18 --recode ped --out lifted --exclude to_exclude.dat
plink --ped lifted.ped --map good_lifted.map --make-bed --out final

# Create files for imputation
plink --bfile final --freq --out final
perl HRC-1000G-check-bim-v4.2.pl -b final.bim -f final.frq -r ~HRC.r1-1.GRCh37.wgs.mac5.sites.tab â€“h

## Execute first lines of Run_plink.sh
## Create files for imputation
for i in $(seq 22)
do
plink --bfile final-updated --chr $i --chr-output M --set-hh-missing --recode vcf-iid --out Alz-chr$i
vcf-sort Alz-chr$i.vcf | bgzip -c > Alz-chr$i.recode.vcf.gz
done
```

Required:

* From github: liftOverPlink.py and rmBadLifts.py
* From Michigan I.S.: HRC.r1-1.GRCh37.wgs.mac5.sites.tab and HRC-1000G-check-bim-v4.2.pl
* From UCSC: liftOver


```R
wget https://raw.githubusercontent.com/sritchie73/liftOverPlink/master/liftOverPlink.py
wget https://raw.githubusercontent.com/sritchie73/liftOverPlink/master/rmBadLifts.py
wget https://github.com/zhanxw/mycode/blob/master/liftOver/hg18ToHg19.over.chain.gz
wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/liftOver
```

## Post imputation

Add ID (from http://www.haplotype-reference-consortium.org/site) and transform to plink binary format (bcftools required: http://www.htslib.org/download/):

```R
#tabix -p vcf HRC.r1-1.GRCh37.wgs.mac5.sites.tab.gz
./bcftools annotate --annotations HRC.r1-1.GRCh37.wgs.mac5.sites.tab.gz --columns CHROM,POS,ID,REF,ALT \
--output chr21.dose.idsrs.vcf.gz --output-type z chr21.dose.vcf.gz

plink --vcf chr21.dose.idsrs.vcf.gz --maf 0.05 --make-bed --out chr21

#remove non-unique
cut -f 2 chr21.bim | sort | uniq -d > dupli_snp.snplist 
plink --bfile chr21 --exclude dupli_snp.snplist  --make-bed --out chr21

####merge chromosomes
plink --merge-list merge_file.txt --out ids
```

