
# Data pre-processing

**Contents**
- Packages used and Data loading
- Quality control at SNP's level
- Quality control at individual's level
- snpMatrix to data.frame

## Packages used and Data loading

Packages used:

```R
library(snpStats)
library(SNPRelate)
library(SNPassoc)
```

This code will show using a LD threshold of 0.2. However, for the project, LD 0.4 and 0.6 were also performed.

Load the data:


```R
al.plink <- read.plink("merged")

geno <- al.plink$genotypes
annot <- al.plink$map
fam <- al.plink$fam

pheno <- read.delim("pheno_LOAD610K.txt", comment.char="#")
rownames(pheno) <- pheno$SUBJ_NO
```

Make sure that the samples are the same for phenotypic and genotypic data:

```R
identical(rownames(geno), rownames(pheno))
ids <- intersect(rownames(pheno), rownames(geno))
geno <- geno[ids,]
pheno <- pheno[ids,]
identical(rownames(geno), rownames(pheno))
fam <- fam[ids,]
```

## Quality control at SNP's level

SNPs filtered:

* Missing data,
* SNPs with a call rate lower than 95% 
* Markers with the minor allele frequency lower than 5% and,
* SNPs not following Hardy-Weinberg equilibrium.

```R
info.snps <- col.summary(geno)
control <- pheno$Case_Control == 0 & !is.na(pheno$Case_Control)
geno.control <- geno[control,]
info.control <- col.summary(geno.control)
use <- info.snps$Call.rate > 0.95 &
  info.snps$MAF > 0.05 &
  abs(info.control$z.HWE < 3.3)
mask.snps <- use & !is.na(use)

geno.qc.snps <- geno[,mask.snps]
annot.qc.snps <- annot[mask.snps,]
```

## Quality control at individual's level

Sex discrepancies (not performed as there was not chr23 in the data):

```R
#get the call rate and the proportion of SNPs heterozygous:
info.indiv <- row.summary(geno.qc.snps)

#Infer gender by X chromosome:
geno.X <- geno.qc.snps[, annot$chromosome == "23" &
!is.na(annot$chromosome)]

info.X <- row.summary(geno.X)

#Filter sex discrepancies by X heterozygosity criteria:
sex.discrep <- (pheno$sex == "male" & info.X$Heterozygosity > 0.2 |
pheno$sex == "female" & info.X$Heterozygosity < 0.2)
```

Outlying heterozygosity using F-statistic:

```R
MAF <- col.summary(geno.qc.snps)$MAF
callmatrix <- !is.na(geno.qc.snps)
hetExp <- callmatrix %*% (2*MAF*(1-MAF))
hetObs <- with(info.indiv, Heterozygosity * (ncol(geno.qc.snps))*Call.rate)
info.indiv$hetF <- 1 - (hetObs/hetExp)
```

Identity-by-descent (IBD) analysis using a linkage disequilibrium (LD) threshold of 0.2 (0.4 and 0.6 were also performed):

```R
snpgdsBED2GDS("merged.bed",
              "merged.fam",
              "merged.bim",
              out = "obGDS")
genofile <- snpgdsOpen("obGDS")
set.seed(12345)
snp.prune <- snpgdsLDpruning(genofile, ld.threshold = 0.2,#0.4, 0.6
                             snp.id = colnames(geno.qc.snps))
snps.ibd <- unlist(snp.prune, use.names = FALSE)

geno.ld <- geno.qc.snps[,snps.ibd]
identical(rownames(pheno), rownames(geno.ld))
```

Remove the individuals:

```R
use <- info.indiv$Call.rate > 0.95 &
abs(info.indiv$hetF) < 0.1 &
!sex.discrep

mask.indiv <- use & !is.na(use)
geno.qc <- geno.ld[mask.indiv,]
pheno.qc <- pheno[mask.indiv,]
identical(rownames(pheno.qc), rownames(geno.qc))
```

Remove cases with no phenotypic information:

```R
pheno <- pheno.qc[complete.cases(pheno.qc$Case_Control),]
geno <- geno.qc[rownames(pheno.qc),]
identical(rownames(pheno), rownames(geno))
summary(as.factor(pheno$Case_Control))
```

## snpMatrix to data.frame

```R
dd <- as.data.frame(geno)

transf <- function(x) {
  new <- as.numeric(x) - 1
  new[new < 0] <- NA
  return(new) }

dd.num <- data.frame(lapply(dd, transf))

#merge genotype with phenotype data
data <- data.frame(cascon = pheno$Case_Control, dd.num)
data$cascon <- as.factor(data$cascon)
rownames(data) <- rownames(pheno)
```