# Data preprocessing

**Contents**
1. Load libraries and data
2. Add inversion genotypes to datasets and discard non-Caucasian samples
3. Select genes and CpG sites from inversion regions
4. Save preprocessed data

## Load libraries and data
Load the required libraries:
```r
library(SummarizedExperiment)
library(GenomicRanges)
library(rexposome)
library(minfi)
library(stringr)
library(SNPassoc)
```
Load data:
```r
load("methylome.Rdata")
load("transcriptome.Rdata")
load("proteome.Rdata")
load("metabolome_serum.Rdata")
load("metabolome_urine.Rdata")
load("imputated_pregnancy_exposome.Rdata")
load("imputated_postnatal_exposome.Rdata")
load("inversion_genotypes.Rdata")
load("inversionGR.rda")
```
## Add inversion genotypes to datasets and discard non-Caucasian samples
Obtain only one imputation from exposome datasets:
```r
pregnancy <- toES(imputated_pregnancy_exposome, rid = 1)
postnatal <- toES(imputated_postnatal_exposome, rid = 1)
```
Transform inversion genotypes to SNPs:
```r
inversions <- as.matrix(inversion_genotypes)
inversions.s <- setupSNP(inversions, 1:ncol(inversions),sep="")
rownames(inversions.s) <- rownames(inversions)
```
Add inversion genotypes to datasets and select Caucasian samples:
```r
add_inv_cauc <- function(dataset){
  inv_ID <- intersect(colnames(dataset),rownames(inversions.s))
  final <- dataset[,inv_ID]
  pData(final) <- cbind(pData(final),inversions.s[inv_ID, ])
  if (class(dataset)!="ExposomeSet"){
    final <- final[, which(final$h_ethnicity_cauc=="yes") ] 
  }
  if (class(dataset)=="ExposomeSet"){
    final <- final[, which(final$h_ethnicity_c_None=="Caucasian") ]
  }
  return(final)
}

methy_final <- add_inv_cauc(methylome)
trans_final <- add_inv_cauc(transcriptome)
prot_final <- add_inv_cauc(proteome)
met_ser_final <- add_inv_cauc(metabolome_serum)
met_uri_final <- add_inv_cauc(metabolome_urine)
preg_final <- add_inv_cauc(pregnancy)
post_final <- add_inv_cauc(postnatal)
```
## Select genes and CpG sites from inversion regions
Create a dataset for each inversion selecting only the features of the inversion region +/- 1Mb.

First, obtain the region of the inversion +/- 1Mb:
```r 
region_inv <- function(inv){
  start <- as.numeric(start(inversionGR[inv,]))-1000000
  end <- as.numeric(end(inversionGR[inv,]))+1000000
  region <- paste(as.character(seqnames(inversionGR[inv,])),":",start,"-",end,sep="")
  return(region)
}
region8 <- GRanges(region_inv("inv8_001"))
region16 <- GRanges(region_inv("inv16_009"))
region17 <- GRanges(region_inv("inv17_007"))
```
Second, create a new methylome dataset for each region: 
```r
methy8 <- subsetByOverlaps(methy_final, region8)
methy16 <- subsetByOverlaps(methy_final, region16)
methy17 <- subsetByOverlaps(methy_final, region17)
```
In the case of transcriptome, we perform another step to convert ExpressionSet to RangedSummarizedExperiment:
```r
 trans_ranged <- makeSummarizedExperimentFromExpressionSet(trans_final)
trans8 <- subsetByOverlaps(trans_ranged, region8)
trans16 <- subsetByOverlaps(trans_ranged, region16)
trans17 <- subsetByOverlaps(trans_ranged, region17)
```

## Save preprocessed data
```r
save(methy_final,file="methy_final.Rdata")
save(trans_final,file="trans_final.Rdata")
save(prot_final,file="prot_final.Rdata")
save(met_ser_final,file="met_ser_final.Rdata")
save(met_uri_final,file="met_uri_final.Rdata")
save(preg_final,file="preg_final.Rdata")
save(post_final,file="post_final.Rdata")
save(methy8,file="methy8.Rdata")
save(methy16,file="methy16.Rdata")
save(methy17,file="methy17.Rdata")
save(trans8,file="trans8.Rdata")
save(trans16,file="trans16.Rdata")
save(trans17,file="trans17.Rdata")
```


