
# Differential analysis

The differential analysis has been performed using transcriptome, methylome, proteome, and metabolome, according to the inversion status of 8p23.1, 16p11.2, and 17q21.31. Furthermore, mean and variance differences has been analysed using three models (dominant, additive, and recessive). 
Here, I show the differential analysis of transcriptome.

Packages used:
```r
library(MEAL)
library(SNPassoc)
library(SummarizedExperiment)
library(dplyr)
```
Load data:
```r
load("trans8.Rdata")
load("trans16.Rdata")
load("trans17.Rdata")
```

Create a function that returns the genes that are differentially expressed with an adjusted p-value bellow 0.0227. 
```r 
diff_inv <- function(dataset,model_inherit,inversion,type_analysis){
  if (type_analysis=="Mean"){
    results <- runDiffMeanAnalysis(set= dataset,model= formula(paste("~",model_inherit,"(",inversion,")+ sex + cohort"),sep=""))
    features <- getProbeResults(results, rid="DiffMean",fNames = NULL)
    topfeatures<-features[ which (features$adj.P.Val < 0.05/2.2), ]
  }
  if (type_analysis=="Variance"){
    results <- runDiffVarAnalysis(set = dataset,model = formula(paste("~",model_inherit,"(",inversion,")+ sex + cohort"),sep=""))
    features <- getProbeResults(results, rid="DiffVar",fNames = NULL)
    topfeatures<-features[ which (features$Adj.P.Val < 0.05/2.2), ]
  }
  return(topfeatures)
}
```
Create a function that collects all the genes differentially expressed in a data.frame:
```r
get_sig_genes <- function(dataset,type,inversion,model_inherit){
  topgenes <- diff_inv(dataset,model_inherit,inversion,type)
  for (transcript in rownames(topgenes)){
    if (type=="Mean"){
      adjpvalue <- "adj.P.Val"
      difflog <- "logFC"
    }
    if (type=="Variance"){
      adjpvalue <- "Adj.P.Value"
      difflog <- "DiffLevene"
    }
    sig_genes_trans_ex <- data.frame(Inversion=inversionGR[inversion,]$Cytogenetic.location,
                                     Type_analysis=type,
                                     Transcript=transcript,
                                     Location=paste(as.character(seqnames(dataset[transcript,])),":",as.character(start(dataset[transcript,])),"-",as.character(end(dataset[transcript,])),sep=""),
                                     Gene_Symbol=allsymbols_to_symbol(paste(rowData(dataset)[transcript,"GeneSymbolDB"],rowData(dataset)[transcript,"GeneSymbolDB2"],sep=";")),
                                     model=model_inherit,
                                     adj.p.value=topgenes[transcript,adjpvalue],
                                     DiffLevene_logFC=topgenes[transcript,difflog])
    sig_genes_trans <- rbind(sig_genes_trans,sig_genes_trans_ex)
  }
  sig_genes_trans <- sig_genes_trans[which(sig_genes_trans$Gene_Symbol!=""),]
  return(sig_genes_trans)
}
```
Create an empty data.frame to store the results:
```r
sig_genes_trans <- data.frame(Inversion=character(),
                              Type_analysis=character(),
                              Transcript=character(),
                              Location=character(),
                              Gene_Symbol=character(),
                              model=character(),
                              adj.p.value=numeric(),
                              DiffLevene_logFC=numeric(),
                              stringsAsFactors=FALSE)
```
Perform the analysis for the three models of inheritande, at variance and mean scale, and for the three most well-studied inversions:
```r
for(model in c("additive","dominant","recessive")){
  for (type in c("Mean","Variance")){
    sig_genes_trans <- get_sig_genes(trans8,type,"inv8_001",model)
    sig_genes_trans <- get_sig_genes(trans16,type,"inv16_009",model)
    sig_genes_trans <- get_sig_genes(trans17,type,"inv17_007",model)
  }
}
```
Finally, store the data.frame with the results:
```r
save(sig_genes_trans,file="sig_genes_trans.Rdata")
```
