
# Enrichment analysis

Again, I am going to show here only the enrichment analysis for the genes differentially expressed in transcriptome as an example. 

Load the required libraries: 
```r
library(clusterProfiler)
library(org.Hs.eg.db)
library(enrichplot)
library(ggplot2)
library(ggupset)
```

Load the DisGeNET database:
```r
gda <- read.delim("curated_gene_disease_associations.tsv.gz")
disease2gene <- gda[, c("diseaseId", "geneId")]
disease2name <- gda[, c("diseaseId", "diseaseName")]
```

Load the top genes for the transcriptome results.
```r
load("sig_genes_trans.Rdata")
```

For instance, the variance differentially expressed genes according to 8p23.1 status are going to be selected for the enrichment analysis:
```r
symbols_trans_v8 <- unique(sig_genes_trans[which(sig_genes_trans$Inversion=="8p23.1" & sig_genes_trans$Type_analysis=="Variance"),]$Gene_Symbol)

First, we need to transform gene symbols to Entrez and Ensembl IDs:
```r
symbolToentrez_ensembl<-function(symbols,entrez_ensemble) {
  symbols<-paste(symbols, collapse=";")
  symbols<-unique(strsplit(symbols,";")[[1]])
  genes <- bitr(symbols, fromType = "SYMBOL",
                toType = entrez_ensemble,
                OrgDb = "org.Hs.eg.db")
  return (genes)
}
entrez_trans_v8 <- symbolToentrez_ensembl(symbols_trans_v8,"ENTREZID")
ensembl_trans_v8 <- symbolToentrez_ensembl(symbols_trans_v8,"ENSEMBL")
```
The next step is to perform the enrichment analysis using these IDs:
* Enrichment using KEGG
```r
kegg_trans_v8  <- enrichKEGG(gene = entrez_trans_v8$ENTREZID,
                         organism = 'hsa',
                         pvalueCutoff = 0.05)
tab_kegg_trans_v8 <- as.data.frame(kegg_trans_v8)
tab_kegg_trans_v8 
p1 <- dotplot(kegg_trans_v8, showCategory=20) + ggtitle("VDE genes for inversion 8p23.1 (KEGG enrichment)")
```
* Enrichment using DisGeNet
```r
dis_trans_v8 <- enricher(entrez_trans_v8$ENTREZID, TERM2GENE=disease2gene,
                      TERM2NAME=disease2name)
tab_dis_trans_v8 <- as.data.frame(dis_trans_v8)
tab_dis_trans_v8 
p2 <- dotplot(dis_trans_v8, showCategory=20) + ggtitle("VDE genes for inversion 8p23.1 (DisGeNET enrichment)")
```
* Enrichment using GO
```r
go_trans_v8 <- enrichGO(gene = ensembl_trans_v8$ENSEMBL, ont = "BP",
                     OrgDb ="org.Hs.eg.db",
                     keyType = "ENSEMBL",
                     readable=TRUE,
                     pvalueCutoff = 0.05)
tab_go_trans_v8 <- as.data.frame(go_trans_v8)
tab_go_trans_v8[1:6,] 
p3 <- dotplot(go_trans_v8, showCategory=20) + ggtitle("VDE genes for inversion 8p23.1 (GO enrichment)")
```

