
# Differential analysis genotype-environment interaction

The methodology for the differential analysis using the interaction between inversion genotype ans pregnancy exposome is the same as the previous differential analysis but modifying some things.

Packages used:
```r
library(MEAL)
library(SNPassoc)
library(SummarizedExperiment)
library(dplyr)
```
Load data:
```r
load("trans8.Rdata")
load("trans16.Rdata")
load("trans17.Rdata")
load("preg_final.Rdata")
```

Create the data sets with the common samples between transcriptome and exposome, and add the exposome data in the transcriptome data set:
```r
intersect_samples<-intersect(colnames(trans8),colnames(preg_final))
colData(trans8)<-cbind(colData(trans8[,intersect_samples]),expos(preg_final[,intersect_samples]))
colData(trans16)<-cbind(colData(trans16[,intersect_samples]),expos(preg_final[,intersect_samples]))
colData(trans17)<-cbind(colData(trans17[,intersect_samples]),expos(preg_final[,intersect_samples]))
exp_variables_preg<-colnames(expos(preg_final))
```
Define the function that returns the genes differentially expressed with an adjusted p-value below 0.05:
```r
diff_inter <- function(dataset, type, inversion, expo){
  if (expo=="h_ln_cat_preg_None" & (inversion=="inv16_009" | inversion=="inv17_007")){
    return(inter_trans)
  }
  topgenes <- diff_inv_expo(dataset, type, inversion, expo)
  for (transcript in rownames(topgenes)){
    if (type=="Mean"){
      adjpvalue <- "adj.P.Val"
      difflog <- "logFC"
    }
    if (type=="Variance"){
      adjpvalue <- "Adj.P.Value"
      difflog <- "DiffLevene"
    }
    df <- data.frame(Exposure=expo,
                     Exposure_abrev=exp_abrev[expo,][["Label for tables"]],
                     Family=exp_abrev[expo,][["Group"]],
                     Inversion=inversionGR[inversion,]$Cytogenetic.location,
                     Type_analysis=type,
                     Transcript = transcript,
                     Location=paste(as.character(seqnames(dataset[transcript,])),":",as.character(start(dataset[transcript,])),"-",as.character(end(dataset[transcript,])),sep=""),
                     Gene_Symbol=allsymbols_to_symbol(paste(rowData(dataset)[transcript,"GeneSymbolDB"],rowData(dataset)[transcript,"GeneSymbolDB2"],sep =";")),
                     adj.p.value=topgenes[transcript,adjpvalue],
                     DiffLevene_logFC=topgenes[transcript,difflog])
                       
    inter_trans <- rbind(inter_trans,df)
  }
  inter_trans <- inter_trans[which(inter_trans$Gene_Symbol!=""),]
  return(inter_trans)
}
```
Create an empty data frame:
```r
inter_trans <- data.frame(Exposure=character(),
                          Exposure_abrev=character(),
                          Family=character(),
                          Inversion=character(),
                          Type_analysis=character(),
                          Transcript=character(),
                          Location=character(),
                          Gene_Symbol=character(),
                          adj.p.value=numeric(),
                          DiffLevene_logFC=numeric(),
                          stringsAsFactors=FALSE)
```
Perform the analysis for the threemost well-studied inversions at variance and mean scale using all the pregnancy exposures:

```r
for (expo in exp_variables_preg){
  for (type in c("Mean","Variance")){
    inter_trans <- diff_inter(trans8, type, "inv8_001", expo)
    inter_trans <- diff_inter(trans16, type, "inv16_009", expo)
    inter_trans <- diff_inter(trans17, type, "inv17_007", expo)
  }
}
```
Save the data frame:
```r
save(inter_trans,file="inter_trans.Rdata")
```
The enrichment analysis of these genes is carried out using the methodology explained in the *Enrichment analysis* markdown file.
