
# Mediation analysis between genomic inversions and gene expression using methylation CpG as mediators

Load libraries:
```r
library(SummarizedExperiment)
library(MultiMed)
library(mediation)
library(SNPassoc)
```
Load data:
```r
load("trans8.Rdata")
load("trans16.Rdata")
load("trans17.Rdata")
load("methy8.Rdata")
load("methy16.Rdata")
load("methy17.Rdata")
load("sig_genes_trans.Rdata")
load("sig_cpgs_methy.Rdata")
```
```r
inv_ID <- intersect(colnames(trans8),colnames(methy8))
inv_genotypes <- colData(trans8)[inv_ID,c("inv8_001","inv16_009","inv17_007","cohort","sex")]
```
Create the function that generated a data set with the transcripts, cpgs and inversions needed for the mediation analysis.
```r
create_medi_dataset <- function(topgenes,topcpgs,trans_dataset,methy_dataset){
  trans_medi <- assays(trans_dataset[as.character(topgenes$Transcript),inv_ID])[["exprs"]]
  methy_medi <- assays(methy_dataset[as.character(topcpgs$CpG),inv_ID])[["Beta"]]
  medi_dataset <- cbind(t(trans_medi),t(methy_medi),inv_genotypes)
  return(medi_dataset)
}
```
Create an empty data frame to store the results:
```r
inv_cpg_gene_mean <- data.frame(Inversion=character(),
                 Transcript=character(),
                 Location.Transcript=character(),
                 Symbol.Transcript=character(),
                 CpG=character(),
                 Location.CpG=character(),
                 Symbol.CpG=character(),
                 p.value.mediation=numeric(),
                 stringsAsFactors = FALSE)
```
Create a function to perform the mediation analysis:
```r
multimed <- function(medi_dataset,inversion,topgenes,topcpgs){
  medi_genes <- colnames(medi_dataset)[grep("^TC", colnames(medi_dataset))]
  medi_cpgs <- colnames(medi_dataset)[grep("^cg", colnames(medi_dataset))]
  for (gene in medi_genes){
    E <- as.numeric(medi_dataset[[inversion]])
    M <- medi_dataset[,medi_cpgs]
    Y <- medi_dataset[[gene]]
    a <- medTest(E, M, Y, nperm = 1000)
    rownames(a) <- medi_cpgs
    if (nrow(a)!=0){
      for (i in 1:nrow(a)){
        de <- data.frame(Inversion=inversionGR[inversion,]$Cytogenetic.location,
                         Transcript=gene,
                         Location.Transcript=topgenes[which(topgenes$Transcript==gene),"Location"],
                         Symbol.Transcript=topgenes[which(topgenes$Transcript==gene),"Gene_Symbol"],
                         CpG=rownames(a)[i],
                         Location.CpG=topcpgs[which(topcpgs$CpG==rownames(a)[i]),"Location"],
                         Symbol.CpG=topcpgs[which(topcpgs$CpG==rownames(a)[i]),"Gene_Symbol"],
                         p.value.mediation=a[i,"p"])
        inv_cpg_gene_var <- rbind(inv_cpg_gene_var,de)
      }
    }
  }
  return(inv_cpg_gene_var)
}
```
As an example, perform the mediation analysis for inversion 8p23.1:
```r
topmean_trans_8 <- select_top_genes("8p23.1","Mean")
topmean_methy_8 <- select_top_cpgs("8p23.1","Mean")

medi_8 <- create_medi_dataset(topmean_trans_8,topmean_methy_8,trans8,methy8)

inv_cpg_gene_mean <- multimed(medi_8,"inv8_001",topmean_trans_8,topmean_methy_8)
```
Look in detail significant mediations using mediation package:
```r
inv_cpg_gene_mean$Prop.Mediated <- NA
inv_cpg_gene_mean$p.value.prop.mediated <- NA

sig_inv_cpg_gene_mean8 <- as.integer(rownames(inv_cpg_gene_mean[which(inv_cpg_gene_mean$p.value.mediation<0.05 & inv_cpg_gene_mean$Inversion=="8p23.1"),]))

mediation <- function(num,medi_dataset,inversion){
  transcript <- as.character(inv_cpg_gene_var[num,"Transcript"])
  cpg <- as.character(inv_cpg_gene_var[num,"CpG"])
  print(paste(transcript,cpg,sep=" vs "))
  design1 <- formula(paste(transcript,"~ additive(",inversion,") + sex + cohort"))
  mod1 <- glm(design1, data=medi_dataset)
  design.M <- formula(paste(cpg,"~ additive(",inversion,") + sex + cohort"))
  model.M <- glm(design.M, data=medi_dataset)
  design.Y <- formula(paste(transcript,"~ additive(",inversion,") +",cpg," + sex + cohort"))
  model.Y <- glm(design.Y, data=medi_dataset)
  treatment=paste("additive(",inversion,")",sep="")
  res <- mediate(model.M, model.Y, treat=treatment, mediator=cpg)
  print(summary(res))
  inv_cpg_gene_var$Prop.Mediated[num] <- summary(res)$n0
  inv_cpg_gene_var$p.value.prop.mediated[num] <- summary(res)$n0.p
  return(inv_cpg_gene_var)
}

for (i in sig_inv_cpg_gene_mean8){
  inv_cpg_gene_mean <-mediation(i,medi_8,"inv8_001")
}
```
After carring out this methodology for the three inversions, save the results:
```r
save(inv_cpg_gene_mean, file="/inv_cpg_gene_mean.Rdata")
```
All these steps are the same for mean and variance scales except the following function that differs. In this case, the variance differences are evaluated after obtaining the residual values using missMethyl package:
```r
create_medi_dataset <- function(topgenes,topcpgs,trans_dataset,methy_dataset){
  trans_medi <- assays(trans_dataset[as.character(topgenes$Transcript),inv_ID])[["exprs"]]
  expr.resid <- missMethyl::getLeveneResiduals(trans_medi, design=NULL)$data
  methy_medi <- assays(methy_dataset[as.character(topcpgs$CpG),inv_ID])[["Beta"]]
  methy.resid <- missMethyl::getLeveneResiduals(methy_medi, design=NULL)$data
  medi_dataset <- cbind(t(expr.resid),t(methy.resid),inv_genotypes)
  return(medi_dataset)
}
```
