
# Association between CpG and genes differentially expressed and methylated with human traits
Load libraries:
```r
library(SummarizedExperiment)
```
Load functions from the aux_functions script:
```r
source("aux_functions.R")
```
Load data:
```r
load("preg_final.Rdata")
load("trans8.Rdata")
load("trans16.Rdata")
load("trans17.Rdata")
load("methy8.Rdata")
load("methy16.Rdata")
load("methy17.Rdata")
```
Obtain common samples and create the appropiate data sets:
```r
obesity_data <- pData(preg_final)
IDs.trans.obesity <- intersect(colnames(trans8),rownames(obesity_data))
IDs.methy.obesity <- intersect(colnames(methy8),rownames(obesity_data))

neuro_data <- read.csv("neuro_helix_jr.csv")
rownames(neuro_data) <- neuro_data[,2]

IDs.sex.neuro <- intersect(colnames(preg_final),rownames(neuro_data))
neuro_data <- cbind(neuro_data[IDs.sex.neuro,],pData(preg_final)[IDs.sex.neuro,53])

IDs.trans.neuro <- intersect(colnames(trans8),rownames(neuro_data))
IDs.methy.neuro <- intersect(colnames(methy8),rownames(neuro_data))
```
Function to evaluate association gene-phenotype
```r
gene_pheno <- function(genes,phenotypes,inv,trans_data,type_pheno,inter_nointer){
  if (type_pheno=="Obesity-related"){
    datos <- cbind(obesity_data[IDs.trans.obesity,],
                   t(assays(trans_data[as.character(genes$Transcript),IDs.trans.obesity])[["exprs"]]))
  }
  if (type_pheno=="Neurological"){
    datos <- cbind(neuro_data[IDs.trans.neuro,],
                   t(assays(trans_data[as.character(genes$Transcript),IDs.trans.neuro])[["exprs"]]))
  }
  for (i in 1:nrow(genes)){
    for (j in 1:nrow(phenotypes)){
      fit <- lm(formula(paste(phenotypes[j,"Phenotype"]," ~", as.character(genes$Transcript[i])," + sex + cohort")), data=datos)
      df_A <- data.frame(Inversion=inv,
                         Type_analysis="Mean",
                         Feature=as.character(genes$Transcript[i]),
                         Diff=inter_nointer,
                         Location=as.character(genes$Location[i]),
                         Gene_Symbol=as.character(genes$Gene_Symbol[i]),
                         Type_phenotype=type_pheno,
                         Phenotype=phenotypes[j,"Phenotype"],
                         Phenotype_abrev=phenotypes[j,"Phenotype_abrev"],
                         p.value=summary(fit)$coefficients[,4][2])
      feature_pheno_mean <- rbind(feature_pheno_mean,df_A)
    }
  }
  return(feature_pheno_mean)
}
```
Function to evaluate association cpg-phenotype:
```r
cpg_pheno <- function(cpgs,phenotypes,inv,methy_data,type_pheno,inter_nointer){
  if (type_pheno=="Obesity-related"){
    datos <- cbind(obesity_data[IDs.methy.obesity,],
                   t(assays(methy_data[as.character(cpgs$CpG),IDs.methy.obesity])[["Beta"]]))
  }
  if (type_pheno=="Neurological"){
    datos <- cbind(neuro_data[IDs.methy.neuro,],
                   t(assays(methy_data[as.character(cpgs$CpG),IDs.methy.obesity])[["Beta"]]))
  }
  for (i in 1:nrow(cpgs)){
    for (j in 1:nrow(phenotypes)){
      fit <- lm(formula(paste(phenotypes[j,"Phenotype"]," ~", as.character(cpgs$CpG[i]),"+ sex + cohort")), data=datos)
      df_A <- data.frame(Inversion=inv,
                         Type_analysis="Mean",
                         Feature=as.character(cpgs$CpG[i]),
                         Diff=inter_nointer,
                         Location=as.character(cpgs$Location[i]),
                         Gene_Symbol=as.character(cpgs$Gene_Symbol[i]),
                         Type_phenotype=type_pheno,
                         Phenotype=phenotypes[j,"Phenotype"],
                         Phenotype_abrev=phenotypes[j,"Phenotype_abrev"],
                         p.value=summary(fit)$coefficients[,4][2])
      feature_pheno_mean <- rbind(feature_pheno_mean,df_A)
    }
  }
  return(feature_pheno_mean)
}
```
We are going to use inversion at 8p23.1 as an example:
```r
#Obesity-related traits associated with inversions
load("obesity_associations.Rdata")
pheno_obesity8 <- unique(df[which(df$p.value<=0.02272727 & df$Inversion=="8p23.1"),c("Phenotype","Phenotype_abrev")])
pheno_obesity8 <- pheno_obesity8[c(2,4,5,6),]

#Neurological traits associated with inversions
load("neuro_associations.Rdata") 
pheno_neuro8 <- unique(df[which(df$p.value<=0.01 & df$Inversion=="8p23.1"),c("Phenotype","Phenotype_abrev")])

#Genes and CpGs differentially expressed/methylated without interaction
load("/home/isglobal.lan/ncarreras/homews/Results_MEAL_trans/sig_genes_trans.Rdata")
load("/home/isglobal.lan/ncarreras/homews/Results_MEAL_methy/sig_cpgs_methy.Rdata")

genes8 <- select_top_genes("8p23.1","Mean")
CpGs8 <- select_top_cpgs("8p23.1","Mean")

#Genes and CpGs differentially expressed with interaction
load("inter_trans.Rdata")
load("inter_methy.Rdata")

genes8_inter <- inter_trans[which(inter_trans$Inversion=="8p23.1" & inter_trans$Type_analysis=="Mean"),]
genes8_inter <- genes8_inter %>% distinct(Transcript, .keep_all = TRUE)
CpGs8_inter <- inter_methy[which(inter_methy$Inversion=="8p23.1" & inter_methy$Type_analysis=="Mean"),]
CpGs8_inter <- CpGs8_inter %>% distinct(CpG, .keep_all = TRUE)
```
Create an empty data.frame to store all the feature(gen/CpG)-phenotype associations:
```r
feature_pheno_mean <- data.frame(Inversion=character(),
                               Type_analysis=character(),
                               Feature=character(),
                               Diff=character(),
                               Location=character(),
                               Gene_Symbol=character(),
                               Type_phenotype=character(),
                               Phenotype=character(),
                               Phenotype_abrev=character(),
                               p.value=numeric(),
                               stringsAsFactors = FALSE)
```
Prove the association with human traits of all the possible mediators:
```r
feature_pheno_mean <- gene_pheno(genes8,pheno_obesity8,"8p23.1",trans8,"Obesity-related","G")
feature_pheno_mean <- cpg_pheno(CpGs8,pheno_obesity8,"8p23.1",methy8,"Obesity-related","G")
feature_pheno_mean <- gene_pheno(genes8_inter,pheno_obesity8,"8p23.1",trans8,"Obesity-related","G*E")
feature_pheno_mean <- cpg_pheno(CpGs8_inter,pheno_obesity8,"8p23.1",methy8,"Obesity-related","G*E")
feature_pheno_mean <- feature_pheno_mean <- gene_pheno(genes8,pheno_neuro8,"8p11.2",trans8,"Neurological","G")
feature_pheno_mean <- cpg_pheno(CpGs8,pheno_neuro8,"8p11.2",methy8,"Neurological","G")
feature_pheno_mean <- gene_pheno(genes8_inter,pheno_neuro8,"8p11.2",trans8,"Neurological","G*E")
feature_pheno_mean <- cpg_pheno(CpGs8_inter,pheno_neuro8,"8p11.2",methy8,"Neurological","G*E")
```
Save the results:
```r
save(feature_pheno_mean,file="feature_pheno_mean.Rdata")
```
Again, the only difference between this analysis (mean level) and the variance level is that we study the association looking at the residual values.
