## QC Analysis

cd Original_Data

## Firstly we removed the chr23 due to the difference of weight of these SNPs between males and females

awk '{ if ($1 >= 1 && $1 <= 22) print $2 }' SNP_file.bim > snp_1_22.txt

plink --bfile SNP_file --extract snp_1_22.txt --make-bed --out SNP_file

# Now we begin with the QC analysis

cd ../QC

### Step 1 ### Explore missingness per individual and per SNP

plink --bfile ../Original_Data/SNP_file --missing    
# output: plink.imiss and plink.lmiss, these files show respectively the proportion of missing SNPs per individual and the proportion of missing individuals per SNP

# make plots to visualize the missingness results
Rscript --no-save hist_miss.R


# Delete SNPs and individuals with high levels of missingness

# Delete SNPs with missingness>0.2
plink --bfile ../Original_Data/SNP_file --geno 0.2 --make-bed --out SNP_file_2

# Delete individuals with missingness>0.2
plink --bfile SNP_file_2 --mind 0.2 --make-bed --out SNP_file_3

# Delete SNPs with missingness>0.02
plink --bfile SNP_file_3 --geno 0.02 --make-bed --out SNP_file_4

# Delete individuals with missingness>0.02
plink --bfile SNP_file_4 --mind 0.02 --make-bed --out SNP_file_5


### Step 2 ### Explore minor allele frequency (MAF)

# make a plot of the MAF distribution
plink --bfile SNP_file_5 --freq --out MAF_check
Rscript --no-save MAF_check.R

# remove SNPs with a low MAF frequency
plink --bfile SNP_file_5 --maf 0.05 --make-bed --out SNP_file_6

# A conventional MAF threshold for a regular GWAS is between 0.01 or 0.05, depending on sample size.


### Step 3 ### Explore Hardy-Weinberg equilibrium (HWE)

# Delete SNPs which are not in Hardy-Weinberg equilibrium.
# Check distribution HWE p-values of all SNPs

plink --bfile SNP_file_6 --hardy
# Selecting SNPs with HWE p-value below 0.00001, required for one of the two plot generated by the next Rscript, allows to zoom in on strongly deviating SNPs 
awk '{ if ($9 <0.00001) print $0 }' plink.hwe>plinkzoomhwe.hwe
Rscript --no-save hwe.R

# By default the --hwe option in plink only filters for controls.
# Therefore, we use two steps, first we use a stringent HWE threshold for controls followed by a less stringent threshold for the case data
plink --bfile SNP_file_6 --hwe 1e-6 --make-bed --out SNP_file_hwe_filter_step1

# The HWE threshold for the cases filters out only SNPs which deviate extremely from HWE 
# This second HWE step only focusses on cases because in the controls all SNPs with a HWE p-value < hwe 1e-6 were already removed
plink --bfile SNP_file_hwe_filter_step1 --hwe 1e-10 --hwe-all --make-bed --out SNP_file_7


### step 4 ###

# Generate a plot of the distribution of the heterozygosity rate of your subjects.
# And remove individuals with a heterozygosity rate deviating more than 3 sd from the mean

# Checks for heterozygosity are performed on a set of SNPs which are not highly correlated.
# Therefore, to generate a list of non-(highly)correlated SNPs, we exclude high inversion regions (inversion.txt [High LD regions]) and prune the SNPs using the command '--indep-pairwise'
# The parameters '50 5 0.2' stand for the window size, the number of SNPs to shift the window at each step and the multiple correlation coefficient for a SNP being regressed on all other SNPs simultaneously

plink --bfile SNP_file_7 --exclude inversion.txt --range --indep-pairwise 50 5 0.2 --out indepSNP
# Note, don't delete the file indepSNP.prune.in, we will use this file in later steps of the analysis

plink --bfile SNP_file_7 --extract indepSNP.prune.in --het --out R_check
# this files contains the pruned data set

# Plot of the heterozygosity rate distribution
Rscript --no-save check_heterozygosity_rate.R

# The following code generates a list of individuals who deviate more than 3 standard deviations from the heterozygosity rate mean.
Rscript --no-save heterozygosity_outliers_list.R

# output: fail-het-qc.txt
# Adapt the file to make it compatible for PLINK, by removing all quotation marks from the file and selecting only the first two columns.
sed 's/"// g' fail-het-qc.txt | awk '{print$1, $2}'> het_fail_ind.txt

# Remove heterozygosity rate outliers
plink --bfile SNP_file_7 --remove het_fail_ind.txt --make-bed --out SNP_file_8


### step 5 ### Explore cryptic relatedness

# It is essential to check in all datasets for cryptic relatedness
# Assuming a random population sample we have excluded all individuals above the pihat threshold of 0.2.

# Check for relationships between individuals with a pihat > 0.2.
plink --bfile SNP_file_8 --extract indepSNP.prune.in --genome --min 0.2 --out pihat_min0.2

# The following commands will visualize specifically these parent-offspring relations, using the z values.
awk '{ if ($8 >0.9) print $0 }' pihat_min0.2.genome>zoom_pihat.genome

# Make a plot to assess the type of relationship.
Rscript --no-save Relatedness.R

# In a random population we suppose that parent-offspring relations are cryptic relations, but it we know that in the dataset
# are families, trio data for example, we have to analyse them with specific family-based methods and we won't filter by founders.
# Let's continue with a random population

plink --bfile SNP_file_8 --filter-founders --make-bed --out SNP_file_9

# Now we will look again for individuals with a pihat >0.2.
plink --bfile SNP_file_9 --extract indepSNP.prune.in --genome --min 0.2 --out pihat_min0.2_in_founders

# For each pair of 'related' individuals with a pihat > 0.2, we recommend to remove the individual with the lowest call rate. 
plink --bfile SNP_file_9 --missing

Rscript low_call_rate_pihat.R

# Adapt the file to make it compatible for PLINK, by removing all quotation marks
sed 's/"// g' 0.2_low_call_rate_pihat.txt | awk '{print$1, $2}'> 0.2_low_call_rate_pihat_bis.txt

# Delete the individuals with the lowest call rate in 'related' pairs with a pihat > 0.2 
plink --bfile SNP_file_9 --remove 0.2_low_call_rate_pihat_bis.txt --make-bed --out SNP_file_10