**Data pre-processing**
=======================

**Contents**

-   Packages used and Data loading
-   Association analysis
-   P-value subsetting
-   GDS conversion

**Packages used and Data loading**
==================================

Packages used

``` r
# Load libraries

library(snpStats)
library(SNPRelate)
```

Data loading

``` r
# Plink

gera1 <- read.plink("/scratch/jgonzalez/DeepLearning/data/GERA1/GERA_T2D_group_1")$genotypes
gera2 <- read.plink("/scratch/jgonzalez/DeepLearning/data/GERA2/GERA_T2D_group_2")$genotypes

pheno1 <- read.delim("/home/isglobal.lan/jgonzalez/PublicData/STUDY/70KforT2D/pheno/GERA1.sample", sep="")
rownames(pheno1) <- pheno1$ID_1
ids <- intersect(rownames(pheno1), rownames(gera1))
gera1 <- gera1[ids,]
pheno1 <- pheno1[ids,]

pheno2 <- read.delim("/home/isglobal.lan/jgonzalez/PublicData/STUDY/70KforT2D/pheno/GERA2.sample", sep="")
rownames(pheno2) <- pheno2$ID_1
ids <- intersect(rownames(pheno2), rownames(gera2))
gera2 <- gera2[ids,]
pheno2 <- pheno2[ids,]

snps <- intersect(colnames(gera1), colnames(gera2))

gera1 <- gera1[,snps]
gera2 <- gera2[,snps]

geno <- rbind(gera1, gera2)

pheno <- rbind(pheno1, pheno2)
```

**Association analysis**
========================

An association analysis was performed with the GWAS data against the diabetes condition, the diferent PCs present in the pheno files were used to adjust the analysis.

``` r
res <- snp.rhs.tests(AFF ~ PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7, data = pheno, snp.data = geno)
```

**P-value subsetting**
======================

Four p-values were used to subset the SNPs IDs in order to train and evaluate different models.

``` r
sig <- c(1, 0.01, 0.001,0.0001)

ps.adj <- p.value(res)

for (i in 1:length(sig)){
  print(sig[i])
  filename <- paste0("/scratch/jgonzalez/DeepLearning/results/GERA/sigSnp_",sig[i],".txt")
  snp.sig <- res.adj[ps.adj < i & !is.na(ps.adj),]
  write.table(snp.sig, file = filename)
}

snp.sig <- res.adj[ps.adj < sig & !is.na(ps.adj),]

pdf("/scratch/jgonzalez/DeepLearning/results/GERA/res.adj.sig.pdf")
chi2.sig <- chi.squared(snp.sig)
qq.chisq(chi2.sig)
dev.off()
```

**GDS conversion**
==================

PLINK text/binary files are converted to GDS files for faster loading and processing times.

``` r
bed <- "/scratch/jgonzalez/DeepLearning/data/GERA1/GERA_T2D_group_1.bed"
fam <- "/scratch/jgonzalez/DeepLearning/data/GERA1/GERA_T2D_group_1.fam"
bim <- "/scratch/jgonzalez/DeepLearning/data/GERA1/GERA_T2D_group_1.bim"

snpgdsBED2GDS(bed, fam, bim, "gera1.gds")

bed <- "/scratch/jgonzalez/DeepLearning/data/GERA2/GERA_T2D_group_2.bed"
fam <- "/scratch/jgonzalez/DeepLearning/data/GERA2/GERA_T2D_group_2.fam"
bim <- "/scratch/jgonzalez/DeepLearning/data/GERA2/GERA_T2D_group_2.bim"

snpgdsBED2GDS(bed, fam, bim, "gera2.gds")
```
